### PR TITLE
WebGLRenderer: Make outputEncoding to a uniform.

### LINKED ISSUE #19056: Shader recompilation when using sRGB encoding and RTT
##### Description of the problem

Hi there,

I'm working on an application that renders the same scene twice in a single frame, once without post processing and once using EffectsComposer. The use case is basically an editor with a thumbnail preview of the scene with postfx.

This seems to trigger a shader recompilation, as has been discovered in a few other issues*. The advice so far has been to only use EffectsComposer and apply gamma correction as a pass in the effects chain, but this is pretty problematic for our use case as we use line rendering quite a bit for e.g. reference grids so losing hardware AA when using RTT results in unacceptable image quality.

A potential work around is setting the encoding of all render targets in the EffectComposer chain to use sRGB encoding, but I'm not certain what problems this might produce and feels brittle since there is no way to _force_ every pass to use the correct encoding as far as I know. But results should still be correct, since my understanding of using sRGB encoding for a texture is that hardware then performs the appropriate conversions on reads/writes, so shaders can still run code as if the data was linear.

It seems that the necessary conditions to reproduce are:
- sRGB encoding for `renderer.outputEncoding`
- using two different cameras
- RTT (EffectsComposer in this case)

Stepping through the code it seems specifically to do with the [`renderContext.state.lights` version](https://github.com/mrdoob/three.js/issues/14121) being different for each camera, which then triggers a check of the various properties against those when the program was created, which in will fail due to the different renderTargets and their encoding.

https://jsfiddle.net/BruOps/qy6o74pn/9/

* Related:
https://github.com/mrdoob/three.js/issues/18893
https://github.com/mrdoob/three.js/issues/18699
https://github.com/mrdoob/three.js/issues/18355
~https://github.com/mrdoob/three.js/issues/14121~

##### Three.js version

- [x] Dev
- [x] r115

Introduced in r112

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

##### Hardware Requirements (graphics card, VR Device, ...)

None

### PR DESCRIPTION
**Pull request by Mugen87**
Fixed #19056.

This PR makes `outputEncoding` to a uniform. That means when the output encoding changes, recompilations of the shader program are not necessary anymore.

It was necessary to change an existing uniform name in `PMREMGenerator` since `outputEncoding` is now a reserved uniform name of `three.js` shaders.

I've tested the change with the mirror example. When setting `WebGLRenderer.outputEncoding` to `THREE.sRGBEncoding`, there is no performance degradation anymore.

### CODE REVIEW

**Comment by mrdoob** [comment]
Interesting!
Could you add builds to this PR so it's easier to test the change in different devices?

**Comment by Mugen87** [comment]
Done: https://raw.githack.com/Mugen87/three.js/dev50/examples/index.html

**Comment by mrdoob** [comment]
On a Pixel 3a, this example loses ~2fps 🤔
https://raw.githack.com/Mugen87/three.js/dev50/examples/index.html?q=phong#webgl_materials_variations_phong
https://threejs.org/examples/index.html?q=phong#webgl_materials_variations_phong

File: build/three.js [hunk 1]
```diff
@@ -18336,17 +18329,32 @@
...
+				'	else if ( outputEncoding == 3002 ) return LinearToRGBE( value );',
+				'	else if ( outputEncoding == 3003 ) return LinearToLogLuv( value );',
+				'	else if ( outputEncoding == 3004 ) return LinearToRGBM( value, 7.0 );',
+				'	else if ( outputEncoding == 3005 ) return LinearToRGBM( value, 16.0 );',
+				'	else if ( outputEncoding == 3006 ) return LinearToRGBD( value, 256.0 );',
+				'	else if ( outputEncoding == 3007 ) return LinearToGamma( value, float( GAMMA_FACTOR ) );',
+				'	else return LinearToLinear( value );',
+
+				'}',
+
```
Comment (mrdoob) [inline build/three.js:18357]: This should go in `encodings_pars_fragmen.glsl` I guess?

**Comment by WestLangley** [comment]
On an iPhone 6, I see a drop from 58 fps to 38 fps.

File: build/three.js (hunk 1 above)
Comment (Mugen87) [reply build/three.js:18357]: Yep, that does work. However, `PMREMGenerator` needs some adjustments.

This class actually already used a similar approach like the renderer now. Meaning input and output encoding are passed in as uniforms. There might be some room for refactoring/clean up in `PMREMGenerator` if this PR gets merged.

**Comment by Mugen87** [comment]
> On an iPhone 6, I see a drop from 58 fps to 38 fps.

I also see a similar performance drop on a Pixel 1. On an iMac, I have 60 FPS in both demos.

 I guess the additional `if`/`else` statements are a problem for mobile devices...

**Comment by mrdoob** [comment]
Yeah, on iPad Air goes from 32FPS to 25FPS... 😕

**Comment by mrdoob** [comment]
Side question, do we still need `LinearToLogLuv` and `LogLuvToLinear`?

**Comment by Mugen87** [comment]
> Yeah, on iPad Air goes from 32FPS to 25FPS

The problem is that any kind of change that makes the renderer solve #19056 produces a certain amount of new overhead. The single material/multiple programs feature will also require more testing in the renderer in order to determine the correct shader. And there will be additional shader compilations times because the renderer will allocate more programs than before.

However, if this PR feels unsafe, we can stall it and try to figure out a solution for the single material/multiple programs approach. It will be then possible to compare the overhead between both solutions.

**Comment by gkjohnson** [comment]
I'm less familiar with the material recompilation so sorry if this isn't helpful but from the outside it's not clear why toggling between two different define values causes such a big hiccup. Intuitively it would seem to me that once a shader is compiled and uploaded it should be cached and reuse-able by any material that shares the set of defines and shader code. Certainly when a define is changed for the first time the shader must be built and recompiled but once switching back shouldn't it be able to reuse same shader that was compiled the first go around? What's happening that causes the big hiccup?

I guess notionally I would expect something like a hash to be based on the defines object, fragment shader, and vertex shader which can be used in a map to point to the shader handle.

Also FWIW I think it's valuable that defines are used in these cases for the sake of performance.

**Comment by Mugen87** [comment]
> shouldn't it be able to reuse same shader that was compiled the first go around?

This is the exact feature that is mentioned above^^. Meaning a single material can have multiple shader programs.

**Comment by Mugen87** [comment]
Related issue: #15047

**Comment by gkjohnson** [comment]
> This is the exact feature that is mentioned above 😉 .

Yeah I see what you're saying now. I suppose I'm not thinking of it as multiple shaders per material. At the moment is a new shader getting compile for every single material in the scene even if two materials are functionally identical? I'm imagining a separate compiled shader cache per renderer that can be used by any material in the scene -- not per material. I think this separate now, though.

Here's another idea:

Considering defines derived from attributes on `WebGLRenderer` seem to be the culprit here would it make sense to keep separate material caches based on keys from renderer-derived defines? If the renderer is swapping between `srgb` and `linear` encodings frequently then a cache would get created for each. Getting the cache once at the start of the render would be fast and sidestep the overhead you mention in #15047 so an appropriate material could be created and cached for each case. `needsUpdate` would still be required for changes to material defines. As far as I'm aware there are only a couple fields on the renderer that would affect this -- the environment map and output encoding come to mind.

A step further would be to share instances between the two caches if the material does not depend on an environment map or output encoding but I don't think the renderer differentiates between them at the moment so this would probably happen separately, as well.

**Comment by Mugen87** [comment]
> At the moment is a new shader getting compile for every single material in the scene even if two materials are functionally identical?

No, the renderer can already detect functional equivalent materials and then reuse the same shader.

Using some sort of cache mechanism is definitely part of #15047. Depending on how it's going to be implement the resulting overhead might be manageable.

**Comment by Mugen87** [comment]
Closing. The resulting overhead of this approach is just too high. It seems more promising to head for #15047.

**Comment by mrdoob** [comment]
Was a good try though! 🙏

**Comment by mrdoob** [comment]
@elalish This is what I was talking about performance regressions when using uniforms for encoding/tonemapping.

**Comment by elalish** [comment]
@mrdoob I see, thanks! I think I've just spent too much time with modern desktop GPUs where this kind of thing has almost no overhead. It sounds like once #15047 lands, we should be able to remove the uniform from PMREM and do it the same way. Hopefully that'll help with perf on these devices.

**Pull request closed without merging**